### ISSUE TITLE
Admin Installation for All Namespaces: incomplete deployments

### ISSUE THREAD

**Issue by k3nlo** [splunk/splunk-operator#206]
OS: Ubunu 20.01 LTS (WSL)
Client: kubectl v1.19.4
Server: Azure Kubernetes Service (AKS) with Kubernetes v1.18.10 (default)

Hi, 
I'm quite new to using the splunk-operator (current Splunk OEM customer) and had a quick question about the recommended splunk-operator installation pattern. 
Let's says we'll be the ones administering the AKS cluster and we'll be running multiple standalones with our application on the same node to service 2~3 small customers. 
Would you recommend a single "Admin installation" of the splunk-operator at the cluster level or an "Installation By a Regular User" with 1 splunk-operator instance per customer namespace?
 
I've been following the "Admin Installation for All Namespaces" and after applying the splunk-operator-cluster.yaml
and binding the ClusterRole splunk:operator:namespace-manager to the ensuing namespaces (splunk-1), standalones deployments dot not proceed/complete (standalone s1 in the splunk-1 namespace does not get a pod) 

I ended up having to sligthly modify the splunk-operator-cluster.yaml file starting at line 12950 for my test/poc. 
What ended up working was:

- not deploying the splunk-operator in its own namespace 'splunk-operator' but rather leaving it in the 'default' namespace.
- creating the ServiceAccount 'splunk-operator' and the ClusterRole 'splunk:operator:namespace-manager' in the 'default' namespace
- ClusterRoleBinding 'splunk:operator:namespace-manager' (but that removes the need to add the RoleBinding in the ensuing namespaces which I'm not sure if it's desirable.)

Any feedback would be greatly appreciated.
Kenlo.
```
kind: ServiceAccount
metadata:
  name: splunk-operator
  # namespace: splunk-operator
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: splunk:operator:namespace-manager
  # namespace: splunk-operator
rules:
- apiGroups:
  - ""
  resources:
  - services
  - endpoints
  - persistentvolumeclaims
  - configmaps
  - secrets
  - pods
  - pods/exec
  verbs:
  - create
  - delete
  - deletecollection
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apps
  resources:
  - deployments
  - daemonsets
  - replicasets
  - statefulsets
  verbs:
  - create
  - delete
  - deletecollection
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - enterprise.splunk.com
  resources:
  - '*'
  verbs:
  - '*'
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: splunk:operator:namespace-manager
  # namespace: splunk-operator
subjects:
- kind: ServiceAccount
  name: splunk-operator
  namespace: default
roleRef:
  kind: ClusterRole
  name: splunk:operator:namespace-manager
  apiGroup: rbac.authorization.k8s.io
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
    rbac.authorization.k8s.io/aggregate-to-admin: "true"
  name: splunk:operator:resource-manager
rules:
- apiGroups:
    - enterprise.splunk.com
  resources:
    - '*'
  verbs:
    - '*'
- apiGroups:
  - apps
  resources:
  - deployments
  - statefulsets
  verbs:
  - list
  - get
  - watch
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: splunk:operator:resource-manager
subjects:
- kind: ServiceAccount
  name: splunk-operator
  namespace: default
roleRef:
  kind: ClusterRole
  name: splunk:operator:resource-manager
  apiGroup: rbac.authorization.k8s.io
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: splunk-operator
  # namespace: splunk-operator
spec:
  replicas: 1
  selector:
    matchLabels:
      name: splunk-operator
  template:
    metadata:
      labels:
        name: splunk-operator
    spec:
      serviceAccountName: splunk-operator
      containers:
      - name: splunk-operator
        image: docker.io/splunk/splunk-operator:0.2.0
        imagePullPolicy: IfNotPresent
        env:
        - name: WATCH_NAMESPACE
          value:
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: OPERATOR_NAME
          value: "splunk-operator"
        - name: RELATED_IMAGE_SPLUNK_ENTERPRISE
          value: "docker.io/splunk/splunk:8.1.0"
        - name: RELATED_IMAGE_SPLUNK_SPARK
          value: "docker.io/splunk/spark:0.0.2"
```
@gaurav-splunk

**Comment by BGrasnick** [comment]
I had the same problem recently where I wanted to set up an EKS cluster with different Splunk Standalone deployments in different namespaces and followed the admin installation instructions for all namespaces here: https://splunk.github.io/splunk-operator/Install.html#admin-installation-for-all-namespaces

While debugging the problem I found the following error that gets send repeatedly from the splunk-operator:

```
reflector.go:123] pkg/mod/k8s.io/client-go@v0.0.0-20191016111102-bec269661e48/tools/cache/reflector.go:96: Failed to list *v1.Secret: secrets is forbidden: User "system:serviceaccount:splunk-operator:splunk-operator" cannot list resource "secrets" in API group "" at the cluster scope
```

When trying to create a Standalone deployment using the following, no pod gets created.

```
cat <<EOF | kubectl apply -f -
---
apiVersion: v1
kind: Namespace
metadata:
  name: customer-dev
---
kind: RoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: splunk:operator:namespace-manager
  namespace: customer-dev
subjects:
- kind: ServiceAccount
  name: splunk-operator
  namespace: splunk-operator
roleRef:
  kind: ClusterRole
  name: splunk:operator:namespace-manager
  apiGroup: rbac.authorization.k8s.io
---
apiVersion: enterprise.splunk.com/v1beta1
kind: Standalone
metadata:
  name: test
  namespace: customer-dev
EOF
```

The fix for me was to change splunk:operator:namespace-manager from RoleBinding to ClusterRoleBinding and some other according small fixes (starting at line 13092 of splunk-operator-cluster.yaml from https://github.com/splunk/splunk-operator/releases/download/0.2.2/splunk-operator-cluster.yaml):

```
# rest omitted and unchanged

apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: splunk:operator:namespace-manager
  #namespace: splunk-operator
rules:
- apiGroups:
  - ""
  resources:
  - services
  - endpoints
  - persistentvolumeclaims
  - configmaps
  - secrets
  - pods
  - pods/exec
  - serviceaccounts
  verbs:
  - create
  - delete
  - deletecollection
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - apps
  resources:
  - deployments
  - daemonsets
  - replicasets
  - statefulsets
  verbs:
  - create
  - delete
  - deletecollection
  - get
  - list
  - patch
  - update
  - watch
- apiGroups:
  - enterprise.splunk.com
  resources:
  - '*'
  verbs:
  - '*'
---
#kind: RoleBinding
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: splunk:operator:namespace-manager
  #namespace: splunk-operator
subjects:
- kind: ServiceAccount
  name: splunk-operator
  # new
  namespace: splunk-operator
roleRef:
  kind: ClusterRole
  name: splunk:operator:namespace-manager
  apiGroup: rbac.authorization.k8s.io


# rest omitted and unchanged
```

Then I can just create a new namespace and a new Standalone deployment in this namespace and need no RoleBinding. This works really well. Is this the recommended approach? What is wrong with the one in the installation instructions?

Thanks already and best regards!

**Comment by prakashmirji** [comment]
Yes - it looks something like below.

```
kubectl get RoleBinding -n splunk
NAME                                ROLE                                     AGE
hpecp-tenant-11-admin-bhgqm         Role/hpecp-tenant-11-admin-vjz87         36h
hpecp-tenant-11-member-qsn2z        Role/hpecp-tenant-11-member-fjlgz        36h
hpecp-tenant-11-sa-24bf9            Role/hpecp-tenant-11-sa-bgz7b            36h
kiali                               Role/kiali                               36h
splunk:operator:namespace-manager   Role/splunk:operator:namespace-manager   7h18m

kubectl get RoleBinding -n splunk splunk:operator:namespace-manager -o yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"rbac.authorization.k8s.io/v1","kind":"RoleBinding","metadata":{"annotations":{},"name":"splunk:operator:namespace-manager","namespace":"splunk"},"roleRef":{"apiGroup":"rbac.authorization.k8s.io","kind":"Role","name":"splunk:operator:namespace-manager"},"subjects":[{"kind":"ServiceAccount","name":"splunk-operator"}]}
  creationTimestamp: "2021-04-23T10:40:45Z"
  managedFields:
  - apiVersion: rbac.authorization.k8s.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:kubectl.kubernetes.io/last-applied-configuration: {}
      f:roleRef:
        f:apiGroup: {}
        f:kind: {}
        f:name: {}
      f:subjects: {}
    manager: kubectl-client-side-apply
    operation: Update
    time: "2021-04-23T10:40:45Z"
  name: splunk:operator:namespace-manager
  namespace: splunk
  resourceVersion: "438388"
  uid: 8688bedf-7212-4036-aa80-b93312b702f3
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: splunk:operator:namespace-manager
subjects:
- kind: ServiceAccount
  name: splunk-operator
```

**Comment by akondur** [comment]
Raised a [PR](https://github.com/splunk/splunk-operator/pull/383) as a solution for the same.

**Comment by leandropadua** [comment]
Why would the operator try to get secrets even when no resources from enterprise.splunk.com exist in the namespaces? 
It looks like this could be reduced by only try to get secrets when actual splunk resources are found.

**Comment by vivekr-splunk** [comment]
upcoming splunk-operator 1.1.0 release addresses this issue. by default in 1.1.0 splunk operator will run cluster-wide. operator will watch all the namespaces. if user wants specific namespace to be watched they have to set "WATCH_NAMESPACE" field in `splunk-operator-config` in `splunk-operator` namespace.

**Comment by akondur** [comment]
Closing as the issue has been fixed in the latest [1.1.0 release](https://github.com/splunk/splunk-operator/releases/tag/1.1.0).
